File: learn-react/TweetApp/backend/src/routes/User.service.js
// User.service.js

const bcrypt = require('bcrypt');
const jwt = require('jsonwebtoken');
const User = require('./User.model');

const registerUser = async (userData) => {
    try {
        const { username, password, name, email, mobileNumber } = userData;

        // Check if the username is already taken
        const existingUser = await User.findOne({ username });
        if (existingUser) {
            throw new Error('Username is already taken.');
        }

        // Hash the password
        const hashedPassword = await bcrypt.hash(password, 10);

        // Create a new user
        const newUser = new User({
            username,
            password: hashedPassword,
            name, // New field
            email, // New field
            mobileNumber, // New field
        });

        await newUser.save();

        // Create and send a JWT token
        const token = jwt.sign({ userId: newUser._id }, 'your-secret-key', {
            expiresIn: '1h',
        });

        return { message: 'User registered successfully', token };
    } catch (error) {
        throw new Error('An error occurred while registering the user.');
    }
};

const loginUser = async (username, password) => {
    try {
        // Check if the username exists
        const user = await User.findOne({ username });

        if (!user) {
            throw new Error('Invalid username or password');
        }

        // Compare the provided password with the stored hashed password
        const passwordMatch = await bcrypt.compare(password, user.password);

        if (!passwordMatch) {
            throw new Error('Invalid username or password');
        }

        // If the username and password are correct, generate a JWT token
        const token = jwt.sign({ userId: user._id, userName: user.name }, 'your-secret-key', { expiresIn: '1h' });

        return token;
    } catch (error) {
        throw new Error('An error occurred during login');
    }
};

const getAllUsers = async () => {
    try {
        const users = await User.find({}, "-password"); // Exclude the password field
        return users;
    } catch (error) {
        throw new Error('Error fetching users');
    }
};

const updateUserRole = async (userId) => {
    try {
        const user = await User.findById(userId);

        if (!user) {
            throw new Error('User not found');
        }

        user.role = "admin"; // Set the user's role as admin
        await user.save();

        return user;
    } catch (error) {
        throw new Error('Error updating user role');
    }
};

const getUserById = async (userId) => {
    try {
        const user = await User.findById(userId, '-password'); // Exclude the password field

        if (!user) {
            throw new Error('User not found');
        }

        return user;
    } catch (error) {
        throw new Error('An error occurred while fetching the user');
    }
};

const updateUserById = async (userId, userData) => {
    try {
        const { username, name, email, mobileNumber } = userData;

        const user = await User.findById(userId);

        if (!user) {
            throw new Error('User not found');
        }

        // Update user fields
        user.username = username || user.username;
        user.name = name || user.name;
        user.email = email || user.email;
        user.mobileNumber = mobileNumber || user.mobileNumber;

        await user.save();

        return user;
    } catch (error) {
        throw new Error('An error occurred while updating the user');
    }
};

module.exports = {
    registerUser,
    loginUser,
    getAllUsers,
    updateUserRole,
    getUserById,
    updateUserById
};

File: learn-react/TweetApp/backend-spring-boot/src/main/java/com/prem/user/service/UserService.java
package com.prem.user.service;

import com.prem.base.exception.CustomValidationException;
import com.prem.base.util.JwtUtil;
import com.prem.user.dto.UserDto;
import com.prem.user.util.UserUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;
import com.prem.user.repository.UserRepository;
import com.prem.user.entity.User;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;

import java.util.Date;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class UserService {

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private BCryptPasswordEncoder passwordEncoder;

    @Autowired
    private UserUtil userUtil;

    @Autowired
    private JwtUtil jwtUtil;

//    private static final String SECRET_KEY = "your-secret-key";

    public String registerUser(UserDto userData) {
        String username = userData.getUsername();
        String password = userData.getPassword();

        Optional<User> existingUser = userRepository.findByUsername(username);
        if (existingUser.isPresent()) {
            throw new CustomValidationException("Username is already taken.");
        }

        String hashedPassword = passwordEncoder.encode(password);

        User newUser = new User();
        newUser.setUsername(username);
        newUser.setPassword(hashedPassword);
        newUser.setName(userData.getName());
        newUser.setEmail(userData.getEmail());
        newUser.setMobileNumber(userData.getMobileNumber());
        newUser.setRole("user");

        userRepository.save(newUser);

        return jwtUtil.generateToken(newUser);
    }

    public String loginUser(String username, String password) {
        Optional<User> user = userRepository.findByUsername(username);
        if (user.isEmpty() || !passwordEncoder.matches(password, user.get().getPassword())) {
            throw new CustomValidationException("Invalid username or password");
        }

        return jwtUtil.generateToken(user.get());
    }

    public List<UserDto> getAllUsers() {
        return userRepository.findAll()
                .stream()
                .map(user -> {
                    user.setPassword(null); // Exclude password
                    return user;
                })
                .map(userUtil::convertToResponse)
                .collect(Collectors.toList());
    }

    public UserDto updateUserRole(String userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new CustomValidationException("User not found"));

        user.setRole("admin");
        user= userRepository.save(user);
        return userUtil.convertToResponse(user);
    }

    public UserDto getUserById(String userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new CustomValidationException("User not found"));
        user.setPassword(null); // Exclude password
        return userUtil.convertToResponse(user);
    }

    public UserDto updateUserById(String userId, UserDto updatedData){
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new CustomValidationException("User not found"));

        user.setUsername(Optional.ofNullable(updatedData.getUsername()).orElse(user.getUsername()));
        user.setName(Optional.ofNullable(updatedData.getName()).orElse(user.getName()));
        user.setEmail(Optional.ofNullable(updatedData.getEmail()).orElse(user.getEmail()));
        user.setMobileNumber(Optional.ofNullable(updatedData.getMobileNumber()).orElse(user.getMobileNumber()));

        user= userRepository.save(user);
        return userUtil.convertToResponse(user);
    }

//    private String generateToken(User user) {
//        return Jwts.builder()
//                .setSubject(user.getUsername())
//                .claim("userId", user.getId())
//                .claim("userName", user.getUsername())
//                .setIssuedAt(new Date(System.currentTimeMillis()))
////                .setExpiration(new Date(System.currentTimeMillis() + 1000 * 60 * 60 * 10)) // 10 hours expiration
//                .setExpiration(new Date(System.currentTimeMillis() + 1000 * 60 * 1)) // 10 hours expiration
//                .signWith(SignatureAlgorithm.HS512, SECRET_KEY)
//                .compact();
//    }
}
